libp2p/protocols/connectivity/autonatv2/client.nim

```diff
@@ -0,0 +1,202 @@
+# Nim-LibP2P
+# Copyright (c) 2025 Status Research & Development GmbH
+# Licensed under either of
+#  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+#  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+# at your option.
+# This file may not be copied, modified, or distributed except according to
+# those terms.
+
+{.push raises: [].}
+
+import results
+import chronos, chronicles
+import
+  ../../protocol,
+  ../../../switch,
+  ../../../multiaddress,
+  ../../../multicodec,
+  ../../../peerid,
+  ../../../protobuf/minprotobuf,
+  ./types,
+  ./utils
+
+logScope:
+  topics = "libp2p autonat v2 client"
+
+const
+  MaxAcceptedDialDataRequest* = 100 * 1024 # 100 KB
+  MaxDialDataResponsePayload* = 1024
+  DefaultDialBackTimeout* = 5.seconds
+
+type AutonatV2Client* = ref object of LPProtocol
+  dialer*: Dial
+  dialBackTimeout: Duration
+  rng: ref HmacDrbgContext
+  expectedNonces: Table[Nonce, Opt[MultiAddress]]
+
+proc handleDialBack(
+    self: AutonatV2Client, conn: Connection, dialBack: DialBack
+) {.async: (raises: [CancelledError, AutonatV2Error, LPStreamError]).} =
+  debug "Handling DialBack",
+    conn = conn, localAddr = conn.localAddr, observedAddr = conn.observedAddr
+
+  if not self.expectedNonces.hasKey(dialBack.nonce):
+    error "Not expecting this nonce", nonce = dialBack.nonce
+    return
+
+  conn.localAddr.withValue(localAddr):
+    debug "Setting expectedNonces",
+      nonce = dialBack.nonce, localAddr = Opt.some(localAddr)
+    self.expectedNonces[dialBack.nonce] = Opt.some(localAddr)
+  else:
+    error "Unable to get localAddr from connection"
+    return
+
+  trace "Sending DialBackResponse"
+  await conn.writeLp(DialBackResponse(status: DialBackStatus.Ok).encode().buffer)
+
+proc new*(
+    T: typedesc[AutonatV2Client],
+    dialer: Dial,
+    rng: ref HmacDrbgContext,
+    dialBackTimeout: Duration = DefaultDialBackTimeout,
+): T =
+  let client = T(dialer: dialer, rng: rng, dialBackTimeout: dialBackTimeout)
+
+  # handler for DialBack messages
+  proc handleStream(
+      conn: Connection, proto: string
+  ) {.async: (raises: [CancelledError]).} =
+    try:
+      let dialBack = DialBack.decode(initProtoBuffer(await conn.readLp(DialBackLpSize))).valueOr:
+        trace "Unable to decode DialBack"
+        return
+      if not await client.handleDialBack(conn, dialBack).withTimeout(
+        client.dialBackTimeout
+      ):
```

Comment:
nit: if `wait` was used then timeout is handled as `except` statement might be more cleaner